Wrappers/Python/cil/optimisation/functions/SubsetSumFunction.py

```diff
@@ -303,39 +297,366 @@ def memory_reset(self):
+        self.probability_threshold = 1.0/(self.update_frequency*self.num_subsets)
+        self.update_probability = 1.0 
+
+    def gradient(self, x, out=None):
+        """
+        Returns a variance-reduced approximate gradient, defined below.
+        For f = 1/num_subsets \sum_{i=1}^num_subsets f_i, the output is computed as follows:
+            - choose a subset j with function next_subset()
+            - check if full_grad_at_snapshot and snapshot should be updated
+            - if they shouldn't then compute
+                subset_gradient - subset_gradient_at_snapshot + full_grad_at_snapshot
+                where 
+                - subset_grad is the gradient of function number j at current_iterate
+                - subset_grad_at_snapshot is the gradient of function number j at snapshot in memory
+                - full_grad_at_snapshot is the approximation of the gradient of f in memory
+                    computed as full_gradient = 1/num_subsets \sum_{i=1}^num_subsets subset_gradient_{i} (snapshot)
+            - otherwise update full_grad_at_snapshot and snapshot
+            - this gives an unbiased estimator of the gradient
+
+        Reference:
+        Kovalev, Dmitry; Horvath, Samuel; Richtarik, Peter
+        "Don’t Jump Through Hoops and Remove Those Loops:
+        SVRG and Katyusha are Better Without the Outer Loop." 
+        International Conference on algorithmioc Learning Theory. 2020.
+        """
+        if not self.gradients_allocated:
+            self.memory_init(x)         
+
+        # Select the next subset (uniformly at random by default). This generates self.subset_num
+        self.next_subset()
+
+        # In first iteration and if the snapshot image has just been updated, the current iterate and snapshot will be the same
+        # thus tmp2 = 0
+        if self.iter == 0 or self.update_probability < self.probability_threshold:
+            self.tmp2 = 0.0 * self.full_gradient
+       # Otherwise, compute the difference between the subset gradient at current iterate and at the snapshot in memory    
+        else:
+            # Compute new gradient for current subset, store in tmp1
+            # tmp1 = gradient F_{subset_num} (x)
+            self.functions[self.subset_num].gradient(x, out=self.tmp1) 
+
+            # Compute difference between current subset function gradient at current iterate (tmp1) and at snapshot, store in tmp2
+            # tmp2 = gradient F_{subset_num} (x) - gradient F_{subset_num} (snapshot)
+            self.tmp1.axpby(1., -1., self.functions[self.subset_num].gradient(self.snapshot), out=self.tmp2) 
+
+        # Compute the output: tmp2 + full_grad
+        if out is None:
+            ret = 0.0 * self.tmp2
+            self.tmp2.axpby(1., 1., self.full_gradient, out=ret)
+        else:
+            self.tmp2.axpby(1., 1., self.full_gradient, out=out)
+
+        self.iter +=1 
```

Comment:
L-SVRG does not require the iteration counter, so this could be removed (atm kept due to uniformity with SVRG)